### ISSUE TITLE
Invalid url format is accepted in "url"

### ISSUE THREAD

**Issue by ioanarusiczki** [mozilla/addons-linter#4038]
Followup for https://github.com/mozilla/addons-linter/issues/4008#issuecomment-965025174

Submit the attached samples or use the linter locally.

Expected result:
"url": "https://test" is not valid. - A red warning is displayed

Actual result: 
addons pass validation

Notes:
this is reproducible for "url" from "developer" (docs: https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/manifest.json/developer) and also homepage_url

[developer.zip](https://github.com/mozilla/addons-linter/files/7512740/developer.zip)
[homepageurl.zip](https://github.com/mozilla/addons-linter/files/7512742/homepageurl.zip)

**Comment by willdurand** [comment]
It is valid according to the `URL` class:

```js
new URL("https://test/")
// URL { href: "https://test/", origin: "https://test", protocol: "https:", username: "", password: "", host: "test", hostname: "test", port: "", pathname: "/", search: "" }
```

**Comment by Rob--W** [comment]
Why do you view the URL as invalid?

`http://test/` is syntactically not that much different from `https://localhost/`

**Comment by ioanarusiczki** [comment]
This throws an error in dev hub though 

![devhub](https://user-images.githubusercontent.com/33448286/141135178-de9cf649-d250-4ab6-bab7-30772fc70d21.png)

**Comment by Rob--W** [comment]
Ah so the test case is a URL without path component (slash): `https://test`. Firefox accepts that as a URL format.

Technically, the URL has to contain a path component, but since a URL without path is accepted by Firefox, I would accept it in the linter too.

Thoughts?

**Comment by willdurand** [comment]
This seems to be a limitation in addons-server (in django's `URLValidator` to be more precise). cc @diox

**Comment by willdurand** [comment]
More test cases on -dev:

![Screenshot 2021-11-11 at 15 33 56](https://user-images.githubusercontent.com/217628/141315751-bb104808-c0a5-4d78-9844-2b34ebf69a32.png)

![Screenshot 2021-11-11 at 15 34 09](https://user-images.githubusercontent.com/217628/141315773-7e3d9198-d24d-4790-9dbc-f3f7eb9bf8ef.png)

**Comment by Rob--W** [comment]
We also checked `http://test:80/`, which is also rejected apparently.

The stricter validation in the server/frontend may be desirable, to help the developer with inputting a valid URL (a URL that can be resolved globally).
However, it’s a bad experience to pass the linter validation but fail at the next step where the developer URL is rejected (as reported here).

Let’s run a code scan to see how common developer URLs without slash (path component) are. If the usage is minor, we could make the linter stricter to encourage developers to specify a (strictly) valid URL.

The query would be:

1. For every manifest.json file,
2. Check whether the `developer.url` key is present, and check whether the URL is invalid. Potentially by just extracting all values, and then we can filter the URLs later.

The query is already being run as part of https://bugzilla.mozilla.org/show_bug.cgi?id=1739602#c4

**Comment by diox** [comment]
On addons-server, we're using https://github.com/django/django/blob/dfa1145a22042dcf9e504a5a7edd5557e3e0d07c/django/core/validators.py#L66-L92

The trailing slash doesn't matter. What matters is having a fully qualified domain name. So, as far as addons-server is concerned:
- https://test is considered invalid
- https://test/ is considered invalid
- https://test:80 is considered invalid
- https://test:80/ is considered invalid

but:
- https://te.st is considered valid
- https://te.st/ is considered valid
- https://te.st:80 is considered valid
- https://te.st:80/ is considered valid

There is an exception specifically for `localhost`, but you can't use just the tld or just the hostname, even though technically it's a valid URL.

**Comment by eviljeff** [comment]
the django validator behaviour seems correct to me, for the AMO case anyway - a developer homepage or really any other url included as metadata can't be valid with a url like `https://test/` because that's not a valid domain on the internet (though it may be a valid domain in a local intranet).  For the linter 🤷‍♂️ ?

**Comment by willdurand** [comment]
Yeah, I think the "only" issue we have here is:

> The stricter validation in the server/frontend may be desirable, to help the developer with inputting a valid URL (a URL that can be resolved globally). However, it’s a bad experience to pass the linter validation but fail at the next step where the developer URL is rejected (as reported here).

I am thinking that we should be a little more strict when validating URLs in the linter (because of the reasons mentioned in the two comments above)

**Comment by Rob--W** [comment]
We've decided to close this, since we already validate that the URL is a syntactically valid URL, and `http://intentionally.garbage` is not much better than `http://garbagewithoutdot`.

This observed behavior is not a regression, since previously the dev could put garbage as the URL and hit this validation issue in the submission flow again. And we haven't seen reports for that scenario.
